File: src/server/seed.ts
import { sql } from "./db";
import { migrate } from "./migrate";
import { createScenario, upsertBuildings } from "./repository";
import type { BoundsInput, CameraInput, Region } from "../shared/types";

const seedRegion = {
  name: "Seeded Connaught Place",
  north: 28.6342,
  south: 28.6287,
  east: 77.2224,
  west: 77.216,
} satisfies BoundsInput & { name: string };

const seedCameras: CameraInput[] = [
  { camera: "CP-CAM-01", lat: 28.63155, long: 77.21735, orientation_deg: 92, fov_deg: 80, range_m: 130 },
  { camera: "CP-CAM-02", lat: 28.6327, long: 77.2192, orientation_deg: 182, fov_deg: 90, range_m: 120 },
  { camera: "CP-CAM-03", lat: 28.6302, long: 77.22095, orientation_deg: 315, fov_deg: 75, range_m: 110 },
];

const seedBuildings = [
  rectangleBuilding("seed-block-a", [
    [77.2177, 28.63255],
    [77.21855, 28.63255],
    [77.21855, 28.63185],
    [77.2177, 28.63185],
  ]),
  rectangleBuilding("seed-block-b", [
    [77.21945, 28.6318],
    [77.22035, 28.6318],
    [77.22035, 28.63095],
    [77.21945, 28.63095],
  ]),
  rectangleBuilding("seed-block-c", [
    [77.21825, 28.63035],
    [77.21905, 28.63035],
    [77.21905, 28.62965],
    [77.21825, 28.62965],
  ]),
];

export async function seed() {
  await migrate();
  const region = await upsertSeedRegion();
  await upsertBuildings(region.id, seedBuildings);
  await replaceSeedScenario(region.id);
  console.log(`seeded region ${region.id}: ${region.name}`);
}

async function upsertSeedRegion(): Promise<Region> {
  const existing = await sql<Region[]>`
    select id::int, name, north, south, east, west, created_at::text
    from regions
    where name = ${seedRegion.name}
    order by id
    limit 1
  `;

  if (existing[0]) {
    const rows = await sql<Region[]>`
      update regions
      set
        north = ${seedRegion.north},
        south = ${seedRegion.south},
        east = ${seedRegion.east},
        west = ${seedRegion.west},
        bbox = ST_MakeEnvelope(${seedRegion.west}, ${seedRegion.south}, ${seedRegion.east}, ${seedRegion.north}, 4326)
      where id = ${existing[0].id}
      returning id::int, name, north, south, east, west, created_at::text
    `;
    return rows[0]!;
  }

  const rows = await sql<Region[]>`
    insert into regions (name, north, south, east, west, bbox)
    values (
      ${seedRegion.name},
      ${seedRegion.north},
      ${seedRegion.south},
      ${seedRegion.east},
      ${seedRegion.west},
      ST_MakeEnvelope(${seedRegion.west}, ${seedRegion.south}, ${seedRegion.east}, ${seedRegion.north}, 4326)
    )
    returning id::int, name, north, south, east, west, created_at::text
  `;
  return rows[0]!;
}

async function replaceSeedScenario(regionId: number) {
  await sql`
    delete from camera_sets
    where region_id = ${regionId}
      and name = ${"Seeded cameras"}
      and source = ${"db"}
  `;
  await createScenario(regionId, "Seeded cameras", "db", seedCameras);
}

function rectangleBuilding(sourceId: string, corners: Array<[number, number]>) {
  const ring = [...corners, corners[0]!] as Array<[number, number]>;
  return {
    sourceType: "seed",
    sourceId,
    tags: { building: "yes", source: "seed" },
    geometry: {
      type: "MultiPolygon",
      coordinates: [[ring]],
    } satisfies GeoJSON.MultiPolygon,
  };
}

if (import.meta.main) {
  await seed();
  await sql.close();
}

File: blindspot/management/commands/seed.py
from django.core.management.base import BaseCommand

from blindspot.lib.osm import fetch_osm_buildings
from blindspot.lib.repository import create_scenario, get_building_feature_rows, list_regions, upsert_buildings
from blindspot.lib.repository import create_region


SEED_REGIONS = [
    {
        "name": "Seeded Connaught Place",
        "north": 28.6342,
        "south": 28.6287,
        "east": 77.2224,
        "west": 77.216,
        "cameras": [
            {"camera": "CP-CAM-01", "lat": 28.63155, "long": 77.21735, "orientation_deg": 92, "fov_deg": 80, "range_m": 230},
            {"camera": "CP-CAM-02", "lat": 28.6327, "long": 77.2192, "orientation_deg": 182, "fov_deg": 95, "range_m": 210},
            {"camera": "CP-CAM-03", "lat": 28.6302, "long": 77.22095, "orientation_deg": 315, "fov_deg": 85, "range_m": 240},
            {"camera": "CP-CAM-04", "lat": 28.6297, "long": 77.2179, "orientation_deg": 38, "fov_deg": 100, "range_m": 220},
            {"camera": "CP-CAM-05", "lat": 28.63325, "long": 77.22135, "orientation_deg": 230, "fov_deg": 75, "range_m": 260},
        ],
    },
    {
        "name": "Seeded Times Square",
        "north": 40.7606,
        "south": 40.7554,
        "east": -73.9822,
        "west": -73.9894,
        "cameras": [
            {"camera": "TS-CAM-01", "lat": 40.7589, "long": -73.9872, "orientation_deg": 55, "fov_deg": 85, "range_m": 260},
            {"camera": "TS-CAM-02", "lat": 40.7568, "long": -73.9854, "orientation_deg": 25, "fov_deg": 70, "range_m": 240},
            {"camera": "TS-CAM-03", "lat": 40.7598, "long": -73.9845, "orientation_deg": 205, "fov_deg": 90, "range_m": 230},
            {"camera": "TS-CAM-04", "lat": 40.7572, "long": -73.9829, "orientation_deg": 285, "fov_deg": 95, "range_m": 250},
            {"camera": "TS-CAM-05", "lat": 40.7601, "long": -73.9884, "orientation_deg": 122, "fov_deg": 80, "range_m": 280},
        ],
    },
    {
        "name": "Seeded Trafalgar Square",
        "north": 51.5104,
        "south": 51.5066,
        "east": -0.1242,
        "west": -0.1306,
        "cameras": [
            {"camera": "TR-CAM-01", "lat": 51.5082, "long": -0.1294, "orientation_deg": 82, "fov_deg": 85, "range_m": 220},
            {"camera": "TR-CAM-02", "lat": 51.5095, "long": -0.1272, "orientation_deg": 172, "fov_deg": 90, "range_m": 200},
            {"camera": "TR-CAM-03", "lat": 51.5075, "long": -0.1253, "orientation_deg": 306, "fov_deg": 85, "range_m": 230},
            {"camera": "TR-CAM-04", "lat": 51.5089, "long": -0.1248, "orientation_deg": 248, "fov_deg": 75, "range_m": 210},
        ],
    },
]


class Command(BaseCommand):
    help = "Seed demo regions and camera scenarios."

    def handle(self, *args, **options):
        existing = {region["name"]: region for region in list_regions()}
        for seed in SEED_REGIONS:
            region = existing.get(seed["name"]) or create_region(seed["name"], seed)
            if not get_building_feature_rows(region["id"]):
                try:
                    upsert_buildings(region["id"], fetch_osm_buildings(region))
                except Exception as exc:
                    self.stderr.write(f"OSM import skipped for {region['name']}: {exc}")
            create_scenario(region["id"], "Seeded cameras", "db", seed["cameras"])
            self.stdout.write(f"seeded region {region['id']}: {region['name']}")
